### PR TITLE
Task/dev 3691/verify new notarization format

### PR DESCRIPTION
**Pull request by Martin01248**
The notarization format is changing:

- Layer "Chroniql": Finished (live in stage, not promoted to prod)
- Layer "ACC": In progress (Markus is working on that)
- Layer "Notarization Verifier": In review -> I mean this very PR here.

Following fields were added
- objectDataProperties
- references

One field changed
- objectData

For testing a Notarization PDF with the new format is required. ACC(Stage) does not yet produce such PDFs. Since @MarkusWagnerAkarion is currently working on exactly that he would be obviously ideal for that.

Here is some example data
```
        "references": "[{\"_id\":\"63e3d7f00f543d0f9a6d261c\",\"name\":\"incidentCategoryRef\",\"timelineId\":\"63e3d7f00f543d0f9a6d261d\",\"type\":\"INCIDENT_CATEGORY\"},{\"_id\":\"63e3d7fc0f543d0f9a6d261e\",\"name\":\"caseCategoriesRef\",\"timelineId\":\"63e3d7fc0f543d0f9a6d261f\",\"type\":\"WHISTLEBLOWING_CASE_CATEGORY\"}]",

        "objectData": "{\"action\":\"0b9947bdfc64d1ae353dde79d1c03a9614225ab67e28f7664d3a0a9e06a4cd6c\",\"actionDetails\":\"44136fa355b3678a1146ad16f7e8649e94fb4fc21fe77e8310c060f61caaff8a\",\"caseFormValues\":\"6670b30ad5921e93b1622db591e258c4aa38394e9109664d464190798283a4d7\",\"caseId\":\"621a3017fafea84c187b2d4e8f6faef895dfc3c3330a79c507a8c4694215955e\",\"contactFormValues\":\"4f53cda18c2baa0c0354bb5f9a3ecbe5ed12ab4d8e11ba873c2f11161202b945\",\"externalReference\":\"12ae32cb1ec02d01eda3581b127c1fee3b0dc53572ed6baf239721a03d82e126\",\"files\":\"d742a39ab645616929751a73afd59db55921bc4ada804d72571c4b20d5bb6f9a\",\"internalComment\":\"f38d402adfd79d3510fd27d2c61d2503abaa87f319ad4fd70a76346d2aee66c0\",\"label\":\"5a8261b06196173b331d555a989cc3ae8b75a951dd36b5eca1dacb8ae88b1e9e\",\"lang\":\"50eac10e7966a04e139902940a2a8e129a4597574c709cb8d01939ace56ff9c3\",\"measures\":\"63e2bd1c58d1088f0733d8eb81bef59981eb884651836eebb9a198123da9179f\",\"messages\":\"c8666e3754dc8bf70c5241cc889891640fdab92bbf3ef24ebc1832bc8a4c34d8\",\"notifyOnUpdate\":\"fcbcf165908dd18a9e49f7ff27810176db8e9f63b4352213741664245224f8aa\",\"priority\":\"248d71d3d9a0020fec203dae69c7ff752dce825c162ba28bf3304fd3bdfa67a0\",\"pwHash\":\"fdfc68fec9502a2bad096987db255599aeefb25088fb42f9bd0aa62ffc65fdd1\",\"pwSalt\":\"ed6920f1bd0ce1bbc9f8827f22c880c2392a823c38ebb42c7dc67ff3e7ad43fa\",\"status\":\"a8519c42d78eadafc0d8e243adf20d076e5ca3ba6e9fb79d0ad31fcf842dd38e\",\"subject\":\"086917766b8927176998ab3676fa4972d65a9db802aa4a098bdee0cb7cdc6d48\",\"templateOrigin\":\"74234e98afe7498fb5daf1f36ac2d78acc339464f950703b8c019892f982b90b\",\"templateParent\":\"74234e98afe7498fb5daf1f36ac2d78acc339464f950703b8c019892f982b90b\",\"templatePreset\":\"fcbcf165908dd18a9e49f7ff27810176db8e9f63b4352213741664245224f8aa\"}",

        "objectDataProperties": {
            "action": "\"REPORTED\"",
            "actionDetails": "{}",
            "caseFormValues": "[{\"label\":\"Mein erstes Formularfeld\",\"values\":[\"Ich habe beobachtet, wie die Vize-Präsidentin einen Geldkoffer ausgetauscht hat mit ...\"]}]",
            "caseId": "\"pltqi\"",
            "contactFormValues": "[]",
            "externalReference": "\"\"",
            "files": "[{\"contentType\":\"image/png\",\"encKey\":\"ee040f2f4f4772b55ff67d8cc3a7b6add16d3ff5614f1627f6d5dd89f3ddb381\",\"hash\":\"b69123dd2cf71eb40010a18075392cc88c1bdbc3d60751f8435af48693fe083c\",\"id\":\"e7d9b1c5-9463-4da3-b105-32904f6f9f4a\",\"iv\":\"01d98b69e24a0e33788cbcd2692c88bb\",\"name\":\"Beweisfoto.png\",\"scanned\":true,\"size\":382800}]",
            "internalComment": "{\"de\":\"Sofortige Weiterleitung an Beauftragten .....\",\"en\":\"Immediate forwarding to commissioner .....\",\"ja\":\"コミッショナーへの即時転送 ......。\"}",
            "label": "{\"de\":\"Korruptionsverdacht: Vize-Präsidentin.\",\"en\":\"Suspected corruption: vice president.\",\"ja\":\"汚職疑惑：副社長。\"}",
            "lang": "\"en\"",
            "measures": "{\"de\":\"Weiterleitung an Entscheidungsträger ...\"}",
            "messages": "[{\"creationDate\":\"2023-02-24T13:01:21.958Z\",\"senderUserId\":\"63e36dc5e70bb342a60e8de7\",\"text\":\"Danke, wir kümmern uns umgehend darum ...\"}]",
            "notifyOnUpdate": "false",
            "priority": "\"CRITICAL\"",
            "pwHash": "\"4be3c64af4c115c16c199f1877969e101c6bcd8d2a4f1365a20fda1976f926771214121e7f3f3791c2c512c3076e02f86c1f015b873d2744bb2d6842b595c193\"",
            "pwSalt": "\"ef1ddb50c7c50c4ecb38a992a5c51933da17e1faa4b7b72e0e86fdfc1a2b7e121fee69819b65f9ff9638d6d4777f57f79978ec10c9969055fe1094d2d7ad6daa\"",
            "status": "\"CLOSED\"",
            "subject": "\"Korruptionsverdacht: Vize-Präsidentin\"",
            "templateOrigin": "null",
            "templateParent": "null",
            "templatePreset": "false"
        }
```

### CODE REVIEW

File: app/verify.js [hunk 1]
```diff
@@ -100,22 +100,44 @@ function fromHex(h) {
   return decodeURIComponent(escape(s));
 }
 
+export function verifyObjectDataProperties(data) {
+  const props = data.object.objectDataProperties;
+  let errorsFound = false;
+  Object.keys(props).forEach((key) => {
+    const keyAndHash = ['"', key, '":"', hashRaw(props[key]), '"'].join('');
```
Comment (manuelmeindl) [inline app/verify.js:107]: you could write this in one string like this:

`/`"${key}": "${hashRaw(props[key])}"/``

File: app/verify.js [hunk 2]
```diff
@@ -100,22 +100,44 @@ function fromHex(h) {
...
+  if (metaDataAndReferencesStep == null) return false;
+  return hashRaw(data.object.references) === metaDataAndReferencesStep.postfix;
+}
+
 export function verifyMetaData(data) {
-  const metaDataStep = data.hashing.steps[1];
+  const metaDataAndReferencesStep = data.hashing.steps[1];
   const secretStep = data.hashing.steps[0];
-  if (metaDataStep == null || secretStep == null) return false;
+  if (metaDataAndReferencesStep == null || secretStep == null) return false;
```
Comment (manuelmeindl) [inline app/verify.js:134]: I'm not sure if this is on purpose, but comparisons should be done with three === if possible

**Comment by manuelmeindl** [comment]
nice job javascript martin :D

File: app/verify.js [hunk 3]
```diff
@@ -100,22 +100,45 @@ function fromHex(h) {
   return decodeURIComponent(escape(s));
 }
 
+export function verifyObjectDataProperties(data) {
+  const props = data.object.objectDataProperties;
+  let errorsFound = false;
+  Object.keys(props).forEach((key) => {
+    const keyAndHash = `"${key}":"${hashRaw(props[key])}"`;
+    console.log(keyAndHash);
```
Comment (manuelmeindl) [inline app/verify.js:108]: forgotten log :)

**Review by manuelmeindl — CHANGES_REQUESTED** [review]
looking good to me. didn't test it yet. shall I approve it after I tested it with my UI ticket?

**Comment by Martin01248** [comment]
> looking good to me. didn't test it yet. shall I approve it after I tested it with my UI ticket?

Yes, let's do it this way. If you have any questions, feel free to ask me.

In the ideal way following 3 things are done together, ideally directly on sprint change:
 - Notarization Verifier: Merge this PR
 - ACC: Merge PR and promote to PROD
 - Chroniql: Promote to PROD

**Pull request merged**